dot/rpc/subscription/listeners.go

```diff
@@ -85,59 +87,90 @@ func (s *StorageObserver) GetFilter() map[string][]byte {
 // Listen to satisfy Listener interface (but is no longer used by StorageObserver)
 func (s *StorageObserver) Listen() {}
 
+// Stop to satisfy Listener interface (but is no longer used by StorageObserver)
+func (s *StorageObserver) Stop() {}
+
 // BlockListener to handle listening for blocks importedChan
 type BlockListener struct {
 	Channel chan *types.Block
 	wsconn  WSConnAPI
 	ChanID  byte
 	subID   uint
+
+	ctx    context.Context
+	cancel context.CancelFunc
 }
 
 // Listen implementation of Listen interface to listen for importedChan changes
 func (l *BlockListener) Listen() {
-	for block := range l.Channel {
-		if block == nil {
-			continue
-		}
-		head, err := modules.HeaderToJSON(*block.Header)
-		if err != nil {
-			logger.Error("failed to convert header to JSON", "error", err)
-		}
+	l.ctx, l.cancel = context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case <-l.ctx.Done():
+				return
+			case block := <-l.Channel:
+				if block == nil {
+					continue
+				}
```

Comment:
when does the channel get closed?

check if channel is closed:
```suggestion
			case block, ok := <-l.Channel:
				if !ok {
					return
				}
```

Reply:
Check added!